python/pynessie/expression_util.py

```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf-8 -*-
+
+"""The purpose of this module is to provide different functions that can produce CEL Expressions based on given filtering parameters."""
+
+#  Copyright (C) 2020 Dremio
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import List
+from typing import Optional
+
+
+def build_query_expression_for_commit_log_flags(
+    query_expression: str, authors: List[str], committers: List[str], since: Optional[str], until: Optional[str]
+) -> Optional[str]:
+    """Producs a CEL expression to be used for filtering the commit log based on the given filtering parameters."""
+    if query_expression:
+        return query_expression
+    expressions = []
+
+    if authors:
+        expressions.append(_expression_for_commit_log_by_author(authors))
+    if committers:
+        expressions.append(_expression_for_commit_log_by_committer(committers))
+    if since:
+        expressions.append(_expression_for_commit_log_by_commit_time_after(since))
+    if until:
+        expressions.append(_expression_for_commit_log_by_commit_time_before(until))
+
+    if len(expressions) > 0:
+        return _and_join(expressions)
+
+    return None
+
+
+def build_query_expression_for_contents_listing_flags(query_expression: str, types: List[str]) -> Optional[str]:
+    """Producs a CEL expression to be used for filtering the content entries based on the given filtering parameters."""
+    if query_expression:
+        return query_expression
+
+    if types:
+        return _expression_for_entries_by_types(types)
+
+    return None
+
+
+def _expression_for_commit_log_by_author(authors: List[str]) -> str:
+    return __generate_expression(authors, "commit.author=='{}'")
```

Comment:
This one and the other `'{}'` formatted strings need to add CEL-string escaping.
It's unlikely at the moment, but still possible to "ruin" an expression without proper escaping - and the error message would be rather confusing for users.
Maybe also add some lose pattern-checks for the different types (timestamp: `[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}([.][0-9]{1-9})?Z` , committer/author: `[a-zA-Z0-9_.+$@-]+` (or whatever makes sense here)).

Maybe some Python library (`dateutil`?) allows some more flexible way for users to specify timestamps and provide a string format that CEL supports.

[celpy](https://github.com/cloud-custodian/cel-python/blob/master/src/celpy/evaluation.py#L2345) has something to adopt here.

Reply:
agreed. For parsing the `dateutil.parser.parse` method should be able to do all the heavy lifting for dates. If it can't parse the date we assume and error. If it can we write the `datetime` back out as a formatted str

Reply:
yes we definitely need to better check for dates and the CEL strings in general. I was planning to do a bunch of those things in a follow-up PR

Reply:
I think that string escaping is an important issue to fix asap. I would like to see clean date handling too. Neither HAVE to be done in this PR but a quick follow up would be good.

Reply:
I was planning to work on the follow-up right after this PR, but didn't want to make stuff here even more difficult to review. I'll open an issue for that in a bit

Reply:
opened https://github.com/projectnessie/nessie/issues/1481